irispy/new_sji.py

```diff
@@ -136,6 +137,78 @@ def __repr__(self):
+            If False, the current instance is overwritten.
+            Default=False
+
+        force: `bool`
+            If not True, applies (undoes) exposure time correction only if unit
+            doesn't (does) already include inverse time.
+            If True, correction is applied (undone) regardless of unit.  Unit is still
+            adjusted accordingly.
+
+        Returns
+        -------
+        result: `None` or `SJICube`
+            If copy=False, the original SJICube is modified with the exposure
+            time correction applied (undone).
+            If copy=True, a new SJICube is returned with the correction
+            applied (undone).
+
+        """
+        # Get exposure time in seconds and change array's shape so that
+        # it can be broadcast with data and uncertainty arrays.
+        exposure_time_s = u.Quantity(self.extra_coords["EXPOSURE TIME"]["value"], unit='s').value
+        if not isinstance(self.extra_coords["EXPOSURE TIME"]["value"], float):
+            if self.data.ndim == 1:
+                pass
+            elif self.data.ndim == 2:
+                exposure_time_s = exposure_time_s[:, np.newaxis]
+            elif self.data.ndim == 3:
+                exposure_time_s = exposure_time_s[:, np.newaxis, np.newaxis]
+            else:
+                raise ValueError(
+                    "SJICube dimensions must be 2 or 3. Dimensions={0}".format(
+                        self.data.ndim))
+        # Based on value on undo kwarg, apply or remove exposure time correction.
+        if undo is True:
+            new_data_arrays, new_unit = iris_tools.uncalculate_exposure_time_correction(
+                (self.data, self.uncertainty.array), self.unit, exposure_time_s, force=force)
+        else:
+            new_data_arrays, new_unit = iris_tools.calculate_exposure_time_correction(
+                (self.data, self.uncertainty.array), self.unit, exposure_time_s, force=force)
+        # Return new instance of SJICube with correction applied/undone.
+        if copy is True:
+            return SJICube(
+                data=new_data_arrays[0], wcs=self.wcs, uncertainty=new_data_arrays[1],
+                unit=new_unit, meta=self.meta, mask=self.mask, missing_axis=self.missing_axis,
+                extra_coords=convert_extra_coords_dict_to_input_format(self.extra_coords,
+                                                                       self.missing_axis))
+        else:
+            for i in range(self.data.shape[0]):
+                for j in range(self.data.shape[1]):
+                    for k in range(self.data.shape[2]):
+                        if not self.data[i, j, k] == np.nan:
+                            self.data[i, j, k] = new_data_arrays[0][i, j, k]
+                        if not self.uncertainty.array[i, j, k] == np.nan:
+                            self.uncertainty.array[i, j, k] = new_data_arrays[1][i, j, k]
```

Comment:
Yes, I'm not surprised this is taking forever with the number of nested for loops you have.  Regarding the ```Can't set attribute``` error, I believe the ```.data``` attribute, which we inherit from ```astropy.nddata.NDData``` can be operated on but not reset.  Therefore, to simplify things, I would delete this else statement altogether so the function always reinstantiates the ```SJICube``` object as you do in the ```if``` part of the statement.  (The ```if``` line (L197) and the ```copy``` kwarg would have to be deleted.)  Then if the user wants to replace the current instance, they can simply do
```my_cube = my_cube.apply_exposure_time_correction()```
instead of
```my_cube.apply_exposure_time_correction()```
This would make it consistent with ```IRISSpectrogram```.  It's also the standard behaviour exhibited by sunpy objects, i.e. if they have a method that operates on their data, that method returns a new copy of the object with the operation performed, rather than overwriting the existing object.  If there's a good case to do something different we can consider it.  But the "return new copy" behaviour should be the default or only option.

Reply:
Alternatively, there is in fact a simple solution if you do want to keep this copy option.  That is:
```python
if copy:
    pass
else:
    self.data[:] = new_data_arrays[0][:]
    self.uncertainty[:] = new_data_arrays[1][:]

Reply:
Also, if we do keep this copy behaviour it should also be extended to ```IRISSpectrogram``` so it and ```SJICube``` behave consistently.

Reply:
Ok, thanks. I will check all of that tomorrow.